### DISCUSSION TITLE
Text alignment in figure layouts

### DISCUSSION THREAD

**Discussion by DillonHammill** [quarto-dev/quarto-cli#1534] [Q&A]
Hi everyone,

How does one change the alignment of text with a Quarto figure layout? Here is a snippet of a layout containing two images and some text underneath:

```{r}
::: {layout-ncol=2}
![]() # image 1
![]() # image 2

This is the text in the layout but it is center-aligned under the layout in the rendered PDF document.

:::
```
The text in the layout is always center-aligned under the layout when rendered to PDF. Does anyone know how to override this to be left aligned in the rendered PDF?

Thanks in advance for your help.

**Comment by dragonstyle** [comment]
Either of these produce pretty good results for me - do either of these work for you?

```markdown
::: {#fig-test layout-ncol=2 fig-cap="This is the text in the layout but it is center-aligned under the layout in the rendered PDF document."}

![](1.jpg){width="250"}

![](2.jpg){width="250"}

:::
```

or

```markdown
::: {layout-ncol=2}

![](1.jpg){width="250"}

![](2.jpg){width="250"}

:::

This is the text in the layout but it is center-aligned under the layout in the rendered PDF document.
```

**Reply by DillonHammill** [reply]
Thanks @dragonstyle, yeah I tried this approach too. Unfortunately the text is always center-aligned in the rendered PDF but left-aligned in the rendered HTML. I would also like to do some addition formatting on the text so I want to avoid using figure captions if possible to bypass the default formatting there.

**Reply by dragonstyle** [reply]
That seems surprising - in the case where I placed the text outside the layout div, there should be no reason that text should be centered. Are you sure that there isn't something else forcing the text to be centered (e.g. is there other markdown or LaTeX in play that might somehow be affecting this)?

**Reply by DillonHammill** [reply]
Sorry @dragonstyle putting it outside the layout does work, but then the text is not kept with the images as the figure layout floats in the document. I will have to keep the text inside the layout to make sure it is always kept with the images. Unless there is a way to anchor the text to the layout somehow?

**Reply by DillonHammill** [reply]
It would be nice if there was a `layout-halign` attribute similar to `layout-valign`.

**Reply by dragonstyle** [reply]
Ahh, right- was not thinking about that figure floating! I see what is happening - the text appearing in the div is actually being treated as another column and being centered within a minipage. 

This is a challenging one as I think even accomplishing this in a LUA filter will be challenging given that we're generating various LaTeX to process the layout (so injecting text in the correct spot would be tough). I'm not sure that I have a great solution short of using the figure caption, unfortunately. 

Well I did think of one approach, but I'm not thrilled with it :(

````markdown
::: {layout="[[1,1], [1]]"}

![](1.jpg){width="250"}

![](2.jpg){width="250"}

:::{.content-visible when-format="pdf"}
```{=tex}
\raggedright{This is the text in the layout but it is center-aligned under the layout in the rendered PDF document.}
```
:::

:::


:::{.content-visible when-format="html"}
This is the text in the layout but it is center-aligned under the layout in the rendered PDF document.
:::

````

We'd appreciate issues in the `quarto-cli` repo if you have ideas (e.g. `layout-align`) about improving this - both table and figure handling will continue to improve and grow I think...

**Reply by DillonHammill** [reply]
Thanks @dragonstyle, I appreciate your help. Quarto is awesome and I look foward to watching it grow.

**Reply by mcanouil** [reply]
@dragonstyle Me being curious, is it Quarto or Pandoc that add the `{\centering ...}` in the minipage layout? By the way, this is not consistent with HTML formats.
(I wanted to figure out the answer of this one earlier today and realised doing so that the centering was hard written, meaning was not that easy to change)

**Reply by dragonstyle** [reply]
It is Quarto - we are treating that text paragraph as another column in the layout and so placing it centered in a minipage...

**Reply by mcanouil** [reply]
So, `\centering` could be replaced by something parsed from the div cell isn't it? Possibly a `text-align` parameter that could work for both PDF and HTML formats.

**Reply by dragonstyle** [reply]
Yes, I think it could (e.g. we could generate various alignment commands)!

**Reply by fuhrmanator** [reply]
I ran into this problem today. I'll post my workaround below.

**Comment by fuhrmanator** [comment]
I'm using the custom layouts, namely two columns to show an image and some text. I'm doing this, because a markdown table results in odd vertical alignment in the PDF (the text aligns with the bottom of the image, which looks bad when you have multiple lines of text). 

Here's the ref to the documentation: https://quarto.org/docs/authoring/figures.html#vertical-alignment

Here's a MWE:

```markdown
::: {layout="[10,-2,30]" layout-valign="center"}
![](https://i.creativecommons.org/l/by/4.0/88x31.png)

This work is licensed under a [Creative Commons Attribution 4.0 International License](http://creativecommons.org/licenses/by/4.0/).
:::
```

This is great in HTML:

![image](https://user-images.githubusercontent.com/7606540/219105369-952b3cae-f8cc-4c21-8770-8b6b70a65c97.png)

The text is left-aligned, which is what I expect.

However, in the PDF output, the text is centered, which is not what I would expect:

![image](https://user-images.githubusercontent.com/7606540/219105628-30e6cc11-2b70-491d-bde5-99282afad43d.png)

I turned on `keep-tex` in the preamble of my MWE, and I can see that there is the use of `\centering`:

```latex
\begin{minipage}[c]{0.83\linewidth}

{\centering 

This work is licensed under a
\href{http://creativecommons.org/licenses/by/4.0/}{Creative Commons
Attribution 4.0 International License}.

}

\end{minipage}%
```

So, I wonder if there's a better way to do this in LaTeX. Perhaps use `\centering` only if it's an image? This would be more consistent with the HTML output.

Based on how the LaTeX is generated, I hacked a workaround for my MWE which relies on `\justifying` from the `ragged2e` package, if anyone needs a fix. Perhaps there's a simpler way?

```markdown
---
format: 
    pdf:
        keep-tex: true
        header-includes: |
          \usepackage{ragged2e}
        
---

::: {layout="[5,-1,30]" layout-valign="center"}
![](https://i.creativecommons.org/l/by/4.0/88x31.png)

This work is licensed under a [Creative Commons Attribution 4.0 International License](http://creativecommons.org/licenses/by/4.0/).
:::

::: {layout="[5,-1,30]" layout-valign="center"}
![](https://i.creativecommons.org/l/by/4.0/88x31.png)

\justifying \noindent This work is licensed under a [Creative Commons Attribution 4.0 International License](http://creativecommons.org/licenses/by/4.0/).
:::
```

![image](https://user-images.githubusercontent.com/7606540/219106889-2ff70a99-dd82-42e9-9948-d45a9ba7d047.png)

Cheers!

**Reply by kompre** [reply]
Thanks for the workaround @fuhrmanator 

Has been there any more development on this issue?
